styluslabs-write{,-bin}: update, add update script, and refactor

UlyssesZh opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

For the license change of `styluslabs-write-bin`: now the package grabs tarball from GitHub release assets, which can be considered AGPL according to the developer. That tarball is exactly the same as what one can download from http://styluslabs.com/write/write-latest.tar.gz.

The old download URL http://styluslabs.com/write/write300.tar.gz should be considered out-dated (see #281368).

Maybe consider deleting `styluslabs-write-bin` (#437075), but I am keeping it because I cannot seem to use the mouse scroll wheel in the built-from-source package.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [x] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Sigmanificient commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review-gha`](https://github.com/Defelo/nixpkgs-review-gha)

Command: `nixpkgs-review pr 437318`
Commit: [`297fea6d1c43c7be1daeb9f11a6bb04da45c2ab6`](https://github.com/NixOS/nixpkgs/commit/297fea6d1c43c7be1daeb9f11a6bb04da45c2ab6) ([subsequent changes](https://github.com/NixOS/nixpkgs/compare/297fea6d1c43c7be1daeb9f11a6bb04da45c2ab6..pull/437318/head))
Merge: [`21d8ba0e195be5b58b8710ed6232bc06c0a26869`](https://github.com/NixOS/nixpkgs/commit/21d8ba0e195be5b58b8710ed6232bc06c0a26869)

Logs: https://github.com/Sigmanificient/nixpkgs-review-gha/actions/runs/18341587685


---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>styluslabs-write</li>
    <li>styluslabs-write-bin</li>
  </ul>
</details>


---
### `aarch64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>styluslabs-write</li>
  </ul>
</details>

SuperSandro2000 reviewed (commented):
Why do we need the same package twice? Unless we have a pretty good justification, we should drop the bin variant.

Coda-Coda commented:
Hi, some of us are [still experiencing](https://github.com/NixOS/nixpkgs/issues/416488) a segfault with the non `-bin` version, there appears to be an issue more serious than just not being able to scroll.

> > Why do we need the same package twice? Unless we have a pretty good justification, we should drop the bin variant.
> 
> I explained the reason in the PR text. There is a bug in the souce version that I did not find a way to fix. The bin package can be dropped if it is fixed.

Please don't drop the `-bin` package yet.

UlyssesZh commented:
I cannot reproduce it, either, but it is a valid reason of keeping the bin package.

SuperSandro2000 commented on pkgs/by-name/st/styluslabs-write-bin/package.nix:
```diff
@@ -1,80 +1,65 @@
+  passthru.updateScript = ./update.sh;
+
+  meta = {
     homepage = "http://www.styluslabs.com/";
```
```suggestion
    homepage = "https://www.styluslabs.com/";
```

UlyssesZh commented:
I rebased onto master to resolve the merge conflict, and I also added `#include <cstdint>` to solve a strange compilation error (akin to #291419, but I don't know why the problem appeared only just now).

SuperSandro2000 commented:
probably from the gcc15 update

SuperSandro2000 commented on pkgs/by-name/st/styluslabs-write/package.nix:
```diff
@@ -65,27 +66,29 @@ stdenv.mkDerivation (finalAttrs: {
     done
 
-    install -Dm444 scribbleres/linux/Write.desktop -t $out/share/applications
+    install -Dm644 scribbleres/linux/Write.desktop -t $out/share/applications
```
note: this will probably end up as 666 in the nix store anyhow because nix cannot differentiate between user, groups and other permissions.

Eveeifyeve reviewed (commented):
I have the same opinion of this as well: https://github.com/NixOS/nixpkgs/pull/437318#pullrequestreview-3318519909

UlyssesZh committed:
styluslabs-write-bin: 300 -> 2025-01-10; add update script; refactor

Signed-off-by: Ulysses Zhan <UlyssesZhan@gmail.com>

UlyssesZh committed:
styluslabs-write: 2024-10-12 -> 2025-01-10; add update script; refactor

Signed-off-by: Ulysses Zhan <UlyssesZhan@gmail.com>

lukts30 reviewed (approved):
At this point, I think it's better we merge this PR than have it open for another year...

And for the record this is me approving the update to `styluslabs-write-bin`. 
As I did not actually get to run `styluslabs-write`, as that still segfaults for me too.
